File: beveren_health/beveren_health/customize/warehouse_cost_center.py
# Copyright (c) 2026, Beveren Software and contributors
# For license information, please see license.txt

import frappe


def get_warehouse_cost_center(warehouse):
	if not warehouse:
		return None
	return frappe.db.get_value("Warehouse", warehouse, "custom_cost_center")


def _apply_cost_center(doc, warehouse, force=False):
	cost_center = get_warehouse_cost_center(warehouse)
	if not cost_center:
		return

	meta = frappe.get_meta(doc.doctype)
	if meta.has_field("cost_center") and (force or not doc.get("cost_center")):
		doc.cost_center = cost_center

	items = doc.get("items") or []
	if not items:
		return

	item_meta = frappe.get_meta(items[0].doctype)
	if not item_meta.has_field("cost_center"):
		return

	for item in items:
		if force or not item.get("cost_center"):
			item.cost_center = cost_center


def set_cost_center_from_set_warehouse(doc, method=None):
	"""PO / PR / PI / Stock Reconciliation: use set_warehouse."""
	warehouse = doc.get("set_warehouse")
	if not warehouse:
		for item in doc.get("items") or []:
			warehouse = item.get("warehouse")
			if warehouse:
				break
	if warehouse:
		_apply_cost_center(doc, warehouse)


def set_cost_center_from_stock_entry_warehouse(doc, method=None):
	"""Stock Entry: prefer to_warehouse, else from_warehouse; rows use t/s warehouse."""
	header_warehouse = doc.get("to_warehouse") or doc.get("from_warehouse")
	if header_warehouse:
		_apply_cost_center(doc, header_warehouse)

	for item in doc.get("items") or []:
		row_warehouse = item.get("t_warehouse") or item.get("s_warehouse")
		cost_center = get_warehouse_cost_center(row_warehouse)
		if cost_center and not item.get("cost_center"):
			item.cost_center = cost_center

File: beveren_health/public/js/warehouse_cost_center.js
// Copyright (c) 2026, Beveren Software and contributors
// For license information, please see license.txt

frappe.provide("beveren_health.warehouse_cost_center");

beveren_health.warehouse_cost_center.get_cost_center = function (warehouse) {
	if (!warehouse) {
		return Promise.resolve(null);
	}
	return frappe.db
		.get_value("Warehouse", warehouse, "custom_cost_center")
		.then((r) => (r && r.message && r.message.custom_cost_center) || null);
};

/**
 * Set header (and optionally item) cost_center from Warehouse.custom_cost_center.
 */
beveren_health.warehouse_cost_center.set_from_warehouse = function (frm, warehouse, opts) {
	opts = opts || {};
	const header_field = opts.header_field || "cost_center";
	const update_items = opts.update_items !== false;
	const item_field = opts.item_field || "cost_center";

	if (!warehouse || !frm) {
		return;
	}

	beveren_health.warehouse_cost_center.get_cost_center(warehouse).then((cost_center) => {
		if (!cost_center) {
			return;
		}

		if (frm.fields_dict[header_field]) {
			frm.set_value(header_field, cost_center);
		}

		if (update_items && frm.doc.items && frm.doc.items.length) {
			frm.doc.items.forEach((row) => {
				if (frappe.meta.has_field(row.doctype, item_field)) {
					frappe.model.set_value(row.doctype, row.name, item_field, cost_center);
				}
			});
		}
	});
};

/**
 * Set a single child row's cost_center from a warehouse.
 */
beveren_health.warehouse_cost_center.set_row_from_warehouse = function (
	frm,
	cdt,
	cdn,
	warehouse,
	field
) {
	field = field || "cost_center";
	if (!warehouse || !frappe.meta.has_field(cdt, field)) {
		return;
	}

	beveren_health.warehouse_cost_center.get_cost_center(warehouse).then((cost_center) => {
		if (cost_center) {
			frappe.model.set_value(cdt, cdn, field, cost_center);
		}
	});
};
